### PR TITLE
DEV-264: Lilypad Launcher (Mac): Auto-update Launcher

### PR DESCRIPTION
**Pull request by danteruiz**
Ticket - https://highfidelity.atlassian.net/secure/RapidBoard.jspa?rapidView=46&projectKey=DEV&modal=detail&selectedIssue=DEV-264&quickFilter=125&assignee=557058%3A56835251-5116-4703-a5d3-30bccdb21af4

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20579/HighFidelity-Beta-PR15985-c23e047.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20579/HighFidelity-Quest-Beta-PR15985-c23e047.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20579/HighFidelity-Beta-PR15985-c23e047.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20579/HighFidelity-Beta-Interface-PR15985-c23e047.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20583/HighFidelity-Beta-PR15985-98694a7.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20583/HighFidelity-Quest-Beta-PR15985-98694a7.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20583/HighFidelity-Beta-PR15985-98694a7.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20583/HighFidelity-Beta-Interface-PR15985-98694a7.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20610/HighFidelity-Beta-PR15985-2c8162b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20610/HighFidelity-Beta-Interface-PR15985-2c8162b.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20610/HighFidelity-Beta-PR15985-2c8162b.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20610/HighFidelity-Quest-Beta-PR15985-2c8162b.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20616/HighFidelity-Beta-PR15985-7ed13a2.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20616/HighFidelity-Beta-Interface-PR15985-7ed13a2.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20616/HighFidelity-Beta-PR15985-7ed13a2.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20616/HighFidelity-Quest-Beta-PR15985-7ed13a2.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20616/HighFidelity-Beta-PR15985-7ed13a2.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20616/HighFidelity-Beta-Interface-PR15985-7ed13a2.exe

File: launchers/darwin/src/DownloadLauncher.m [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+-(void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didWriteData:(int64_t)bytesWritten totalBytesWritten:(int64_t)totalBytesWritten totalBytesExpectedToWrite:(int64_t)totalBytesExpectedToWrite {
+    CGFloat prog = (float)totalBytesWritten/totalBytesExpectedToWrite;
+    NSLog(@"Launcher downloaded %f", (100.0*prog));
+
+}
+
+-(void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didResumeAtOffset:(int64_t)fileOffset expectedTotalBytes:(int64_t)expectedTotalBytes {
+    // unused in this example
+}
```
Comment (huffman) [inline launchers/darwin/src/DownloadLauncher.m]: Dead code?

File: launchers/darwin/src/updater/main.m [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    for (int index = 0; index < argc; index++) {
+        NSLog(@"argv at index %d = %s", index, argv[index]);
+    }
+    NSString* oldLauncher = [NSString stringWithUTF8String:argv[1]];
+    NSString* newLauncher = [NSString stringWithUTF8String:argv[2]];
+    NSFileManager* fileManager = [NSFileManager defaultManager];
+    [fileManager removeItemAtURL:[UpdaterHelper NSStringToNSURL:oldLauncher] error:nil];
+    [fileManager moveItemAtURL: [UpdaterHelper NSStringToNSURL: newLauncher] toURL: [UpdaterHelper NSStringToNSURL:oldLauncher] error:nil];
```
Comment (huffman) [inline launchers/darwin/src/updater/main.m]: This updater looks good, but I think we can harden this process a little more. By removing a file and then moving a file to that location, there is a window for things to break down - if the updater crashes or stops for any reason, or if the subsequent move fails, the user will be stuck without a launcher, or perhaps with the old launcher (if the remove fails) trying to update itself over and over again.

It looks like https://developer.apple.com/documentation/foundation/nsfilemanager/1412432-replaceitematurl?language=objc provides a way to run this update atomically, assuming the URLs are on the same volume. I think we should look into using that here (or a similar method).

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20741/HighFidelity-Beta-PR15985-b9106b9.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20741/HighFidelity-Beta-Interface-PR15985-b9106b9.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20741/HighFidelity-Beta-PR15985-b9106b9.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20741/HighFidelity-Quest-Beta-PR15985-b9106b9.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20741/HighFidelity-Beta-PR15985-b9106b9.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20741/HighFidelity-Beta-Interface-PR15985-b9106b9.exe

**Pull request merged**